route/src/route_generator_worker.cpp

```diff
@@ -438,5 +438,91 @@ namespace route {
         return true;
     }
 
+bool RouteGeneratorWorker::crosstrack_error_check(const geometry_msgs::PoseStampedConstPtr& msg, lanelet::ConstLanelet current, carma_wm::TrackPos llt_track)
+{
+  if (!world_model_->getMap()) 
+  {
+    ROS_ERROR_STREAM("WMListener received a route before a map was available. Dropping route message.");
+    return true;
+  }
+
+   
+
+  auto route = lanelet::ConstLanelets();
+  for(auto id : route_msg_.route_path_lanelet_ids)
+  {
+    auto ll = world_model_->getMap()->laneletLayer.get(id);
+    route.push_back(ll);
+  }
+
+  bool out_of_llt_bounds = false;
+
+    lanelet::BasicPoint2d position;
+
+    position.x()= msg->pose.position.x;
+    position.y()= msg->pose.position.y;  
+
+    if (boost::geometry::distance(position, current.polygon2d()) > cross_track_dist) //Evaluate lanelet crosstrack distance from vehicle
+        {
+            cte_count++;
+
+            if(cte_count > 4) //If the distance exceeds the crosstrack distance a certain number of times, report that the route has been departed
+                return true;
+        }
+
+         if (boost::geometry::distance(position, current.polygon2d()) > llt_track.downtrack) //Evaluate lanelet downtrack distance
+        {
+            cte_count++;
+
+            if(cte_count > 4)
+                return true;
+        }
+
+  if(boost::geometry::within(position, current.polygon2d()) == false) //Determine whether or not the vehicle is in the lanelet polygon
+      {
+        out_of_llt_bounds = true;
+      }
+
+    auto following_llts = world_model_->getMapRoutingGraph()->following(current); //Get all subsequent lanelets from the map
+    lanelet::ConstLanelets filtered_llts;
+    for(auto i:following_llts)
+    {
+       for(auto j: route)
+       {
+           //Filter lanelets from following lanelets that are not a part of the route based on id
+           if (i.id() == j.id())
+            filtered_llts.push_back(i);
+       }
+    }
+
+    for(auto i: filtered_llts)
+    {
+        if (boost::geometry::within(position, i.polygon2d()) == false) 
+           out_count++;
+           
+        if (boost::geometry::distance(position, current.polygon2d()) > llt_track.downtrack)
+            out_of_llt_bounds = true;
+    }
+
+    if(out_count == filtered_llts.size()) //If our current position is not within the bounds of any of the following lanelets, then return true
```

Comment:
I think I understand what's being done here, but it may be algorithmically simpler (and faster) to iterate over the list of lanelets in the route, if we find one that we're in, we stop and return "No CTE violation", else if we hit the end of the list we return a CTE violation. The current way we're stepping through every lanelet every timestep it seems.

Reply:
I see your point. I'll modify it to match that logic.